examples/text/plot_document_clustering.py

```diff
@@ -3,284 +3,423 @@
+# labelling, whereas perfect labeling is scored 1.0. Notice from the scores
+# above that the cluster assignment is indeed above chance level, but the
+# overall quality can certainly improve.
+#
+# Performing dimensionality reduction using LSA
+# ---------------------------------------------
+#
+# A `n_init=1` can still be used as long as the dimension of the vectorized
+# space is reduced first to make K-means more stable. For such purpose we use
+# :class:`~sklearn.decomposition.TruncatedSVD`, which works on term count/tf-idf
+# matrices. Since SVD results are not normalized, we redo the normalization to
+# improve the :class:`~sklearn.cluster.KMeans` result. Such process is known as
+# latent semantic analysis (LSA).
+
+from sklearn.decomposition import TruncatedSVD
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import Normalizer
 
-labels = dataset.target
-true_k = np.unique(labels).shape[0]
 
-print("Extracting features from the training dataset using a sparse vectorizer")
+lsa = make_pipeline(TruncatedSVD(n_components=100), Normalizer(copy=False))
 t0 = time()
-if opts.use_hashing:
-    if opts.use_idf:
-        # Perform an IDF normalization on the output of HashingVectorizer
-        hasher = HashingVectorizer(
-            n_features=opts.n_features,
-            stop_words="english",
-            alternate_sign=False,
-            norm=None,
-        )
-        vectorizer = make_pipeline(hasher, TfidfTransformer())
-    else:
-        vectorizer = HashingVectorizer(
-            n_features=opts.n_features,
-            stop_words="english",
-            alternate_sign=False,
-            norm="l2",
-        )
-else:
-    vectorizer = TfidfVectorizer(
-        max_df=0.5,
-        max_features=opts.n_features,
-        min_df=2,
-        stop_words="english",
-        use_idf=opts.use_idf,
-    )
-X = vectorizer.fit_transform(dataset.data)
-
-print("done in %fs" % (time() - t0))
-print("n_samples: %d, n_features: %d" % X.shape)
-print()
+X_lsa = lsa.fit_transform(X_tfidf)
+explained_variance = lsa[0].explained_variance_ratio_.sum()
 
-if opts.n_components:
-    print("Performing dimensionality reduction using LSA")
-    t0 = time()
-    # Since LSA/SVD results are not normalized,
-    # we redo the normalization to improve the k-means result.
-    svd = TruncatedSVD(opts.n_components)
-    normalizer = Normalizer(copy=False)
-    lsa = make_pipeline(svd, normalizer)
+print(f"LSA done in {time() - t0:.3f} s")
+print(f"Explained variance of the SVD step: {explained_variance * 100:.1f}%")
 
-    X = lsa.fit_transform(X)
+# %%
+# Using a single initialization means the processing time will be reduced for
+# both :class:`~sklearn.cluster.KMeans` and
+# :class:`~sklearn.cluster.MiniBatchKMeans`.
+
+kmeans = KMeans(
+    n_clusters=true_k,
+    max_iter=100,
+    n_init=1,
+)
 
-    print("done in %fs" % (time() - t0))
+fit_and_evaluate(kmeans, X_lsa, name="KMeans\nwith LSA on tf-idf vectors")
```

Comment:
This is so fast that this displays:

```
clustering done in 0.0 ± 0.0 s
```

We should probably use 3 digits (milliseconds) to measure the all the durations recorded throughout this example.

Reply:
I changed it to use 2 digits because the uncertainty is sometimes in the first digit. Would you still prefer to use milliseconds for interpretability?